File: task_equivalents/python/machine_learning_workflows/02_regression_workflow_feature_importance.py
# Extracted from Ch08.py (Q10 boosting/lasso workflow)

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoCV
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error

Hitters = pd.read_csv('C:\\Users\\Carol\\Desktop\\Hitters.csv').drop('Unnamed: 0', axis=1).dropna()
Hitters['League'] = Hitters.League.map({'A':0, 'N':1})
Hitters['NewLeague'] = Hitters.NewLeague.map({'A':0, 'N':1})
Hitters['Division'] = Hitters.Division.map({'E':0, 'W':1})
Hitters['logSal'] = np.log(Hitters.Salary)

x = Hitters.drop(['Salary', 'logSal'], axis=1)
y = Hitters['logSal']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=200, random_state=1)

lambda_grid = np.linspace(0.001, 0.201, 20)
train_mse = np.zeros(20)
for i in range(20):
    boost = GradientBoostingRegressor(n_estimators=1000, learning_rate=lambda_grid[i], random_state=2)
    boost.fit(xtrain, ytrain)
    train_mse[i] = mean_squared_error(ytrain, boost.predict(xtrain))

lasso = LassoCV(cv=10, normalize=True).fit(xtrain, ytrain)
lasso_mse = mean_squared_error(ytest, lasso.predict(xtest))

best_lambda = lambda_grid[train_mse.argmin()]
best_boost = GradientBoostingRegressor(n_estimators=1000, learning_rate=best_lambda, random_state=2)
best_boost.fit(xtrain, ytrain)

feature_importance = pd.Series(best_boost.feature_importances_ * 100, index=xtrain.columns).sort_values(ascending=False)
print(feature_importance)

File: task_equivalents/r/machine_learning_workflows/02_regression_workflow_feature_importance.R
# Extracted from Ch08.R (Q10 boosting/lasso workflow)

library(gbm)
library(glmnet)
Hitters = read.csv('data/csv/Hitters.csv', header=TRUE, row.names=1)

Hitters = na.omit(Hitters)
drop_cols = intersect(c('rownames', 'Player', 'Name'), colnames(Hitters))
Hitters = Hitters[, setdiff(colnames(Hitters), drop_cols), drop=FALSE]
Hitters$League = ifelse(Hitters$League == 'A', 0, 1)
Hitters$NewLeague = ifelse(Hitters$NewLeague == 'A', 0, 1)
Hitters$Division = ifelse(Hitters$Division == 'E', 0, 1)
Hitters$logSal = log(Hitters$Salary)

x = subset(Hitters, select=-c(Salary, logSal))
y = Hitters$logSal

set.seed(1)
train = sample(1:nrow(x), 200)
test = setdiff(1:nrow(x), train)
xtrain = x[train,]
xtest = x[test,]
ytrain = y[train]
ytest = y[test]

train.mse = rep(0, 20)
lambda.grid = seq(0.001, 0.201, length.out=20)
for (i in seq_along(lambda.grid)) {
  boost.hit = gbm(ytrain~., data=data.frame(xtrain, ytrain=ytrain), distribution='gaussian', n.trees=1000, shrinkage=lambda.grid[i], verbose=FALSE)
  boost.pred = predict(boost.hit, newdata=xtrain, n.trees=1000)
  train.mse[i] = mean((boost.pred - ytrain)^2)
}

cv.out     = cv.glmnet(as.matrix(xtrain), ytrain, alpha=1, nfolds=10)
bestlam    = cv.out$lambda.min
lasso.pred = predict(cv.out, s=bestlam, newx=as.matrix(xtest))
lasso.mse  = mean((lasso.pred-ytest)^2)

best.lambda = lambda.grid[which.min(train.mse)]
best.boost.hit = gbm(ytrain~., data=data.frame(xtrain, ytrain=ytrain), distribution='gaussian', n.trees=1000, shrinkage=best.lambda, verbose=FALSE)
summary(best.boost.hit)
